spec/models/enterprise_spec.rb

```diff
@@ -195,7 +195,7 @@
       e2 = create(:enterprise, permalink: "not_taken")
       e2.permalink = "taken"
       e2.save
-      expect(e2.permalink).to eq "not_taken"
+      expect(e2.reload.permalink).to eq "not_taken"
```

Comment:
What was the error here? I feel like this implies that there's some breaking changes in the behavior of some of the methods used in that callback, maybe it's worth having a closer look at...

https://github.com/openfoodfoundation/openfoodnetwork/blob/a0476be2c988af10dd4a6711586280e4a0b3815b/app/models/enterprise.rb#L462-L465

Reply:
```
1) Enterprise callbacks restores permalink to original value when it is changed and invalid
Paperclip::Shoulda::Matchers::HaveAttachedFileMatcher implements a legacy RSpec matcher
     Failure/Error: expect(e2.permalink).to eq "not_taken"
expected: "not_taken"
            got: "taken"
     
       (compared using ==)
     # ./spec/models/enterprise_spec.rb:198:in `block (3 levels) in <top (required)>'
```
Yeah it looked similar to the 5.2 changes around callbacks and changes to `attribute_was`, `attribute_changed` etc. I stuck a debugger in there and tested it - `permalink_was` had the correct value of `not_taken`, and `permalink_changed` and `errors[:permalink].present?` were both true, which is what I expected.

Reply:
![Fry](https://user-images.githubusercontent.com/9029026/117578946-20dea080-b0f1-11eb-8fff-e2a2f1671cb6.jpg)